walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java

```diff
@@ -14,10 +14,116 @@
+        if (eventAttendees.contains(attendee)) {
+          attendingEvent = true;
+          break;
+        }
+      }
+      boolean optionals = false;
+      if (attendingEvent == false) {
+        for (String attendee : optionalAttendees) {
+          if (eventAttendees.contains(attendee)) {
+            optionals = true;
+            break;
+          }
+        }
+        if (optionals == false) {
+          continue;
+        }
+      }
+
+      if (optionals == false) {
+        checkEventCompatibility(mandatoryTimes, event.getWhen(), duration);
+      } else {
+        checkEventCompatibility(optionalTimes, event.getWhen(), duration);
+      }
+    }
+
+    // if TimeRanges from optionalTimes and mandatoryTimes overlap, and if there is room
+    // for another TimeRange within the overlaping time, add the TimeRange to
+    // availableTimes
+    ArrayList<TimeRange> availableTimes = new ArrayList<>();
+    for (TimeRange optionalTime : optionalTimes) {
+      for (TimeRange mandatoryTime : mandatoryTimes) {
+        if (optionalTime.overlaps(mandatoryTime)) {
+          int start = optionalTime.start() < mandatoryTime.start() ? mandatoryTime.start()
+                                                                   : optionalTime.start();
+          int end =
+              optionalTime.end() < mandatoryTime.end() ? optionalTime.end() : mandatoryTime.end();
+          if (end == TimeRange.END_OF_DAY && end - start >= duration - 1) {
+            availableTimes.add(TimeRange.fromStartEnd(start, end, true));
+          } else if (end - start >= duration) {
+            availableTimes.add(TimeRange.fromStartEnd(start, end, false));
+          }
+        }
+      }
+    }
+
+    if (availableTimes.isEmpty()) {
+      if (optionalTimes.isEmpty() && attendees.isEmpty()) {
+        return optionalTimes;
+      }
+      return mandatoryTimes;
+    } else {
+      return availableTimes;
+    }
+  }
+
+  // if any TimeRanges in timeRanges overlap with the eventTimeRange, remove the current
+  // TimeRange and add another if there is enough room before or after the eventTimeRange
+  private void checkEventCompatibility(
+      ArrayList<TimeRange> timeRanges, TimeRange eventTimeRange, long duration) {
+    int i = 0;
+    while (i < timeRanges.size()) {
+      TimeRange timeRange = timeRanges.get(i);
+      if (timeRange.overlaps(eventTimeRange)) {
+        if (timeRange.start() + duration <= eventTimeRange.start()) {
+          timeRanges.add(
+              TimeRange.fromStartEnd(timeRange.start(), eventTimeRange.start(), false));
+        }
+        if (timeRange.end() >= duration + eventTimeRange.end()) {
+          if (timeRange.end() == TimeRange.END_OF_DAY) {
+            timeRanges.add(
+                TimeRange.fromStartEnd(eventTimeRange.end(), timeRange.end(), true));
+          } else {
+            timeRanges.add(
+                TimeRange.fromStartEnd(eventTimeRange.end(), timeRange.end(), false));
+          }
```

Comment:
Nit: Can collapse to single call with last arg being "timeRange.end() == TimeRange.END_OF_DAY".